### DISCUSSION TITLE
🚀 FR add a language parameter to date formatting

### DISCUSSION THREAD

**Discussion by robinsowell** [ExpressionEngine/ExpressionEngine#3579] [Feature Requests]
### Suggest an idea for ExpressionEngine

While a bit edge case, folks handle multi-language sites in a variety of ways, often involving passing a variable to indicate language (rather than using the member localization to set it).

So with custom fields, you can easily do something like {lang}_body and have en_body and es_body fields show depending on how you definite language.

Date formatting does use language packs, so it would be nice to have a language parameter available to allow you to override the localized date.  We already give them a timestamp parameter that overrides localization, a language one continues in that vein.

**Comment by PDXMan1** [comment]
My nonprofit and NGO clients serve people from all over the world, so sometimes a website needs to provide translated content to the public, and EE makes 99.9% of that really straightforward. The missing piece is when our clients offer training and informational events that are targeted at non-English speaking audiences. We can't currently translate the date and time info for them. 

Imagine if you read (in your language) about a training event that could help you find stable housing for your family, but you couldn't read the date information to know it happened a week ago, because you couldn't read the month abbreviation.

What would help, is the ability to set a language parameter within EE's date tags so that they could access EE's language packs for languages other than what the CP is using, and/or without requiring anyone to log-in first.

`{current_time lang="{embed:two-character-language-code}" format="%D, %F %d, %Y"}`

<img width="750" alt="Cal1" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/6145078/24e40eb8-f898-4163-a5e5-47adc0050401">

**Comment by intoeetive** [comment]
I think this might be not just about the dates. There are some system messages that are shown based on language files

The dates (and other content) is localized based on member's preferences - however if you're not logged, site's default will be used (which might be not the language that the content is in)

I think we need to make the **system** aware of the language that needs to be used - not just the date parser.

I found some references in the code that tell you can set `exp_language` cookie with language name, and it will be used. However I don't see any code that would actually set that cookie. Might be something from EE1 times - @robinsowell does that ring the bell at all?

So I think we could add a template tag that would allow setting that cookie.

Another option is the ability to declare something like `{lang="es"}` at the top of the template, which will switch the language just for this template parsing. But - not everythig is templates. 

I think I like the cookie option the best, but really open to any other option.

Like, the channel has 'XML Language' setting - we could use that for content that in that channel

**Comment by PDXMan1** [comment]
Please avoid using cookies for the solution as the cookies can get the site stuck in a language sometimes, and our users and visitors need the ability to easily switch between languages on the front end. We tried cookie-based solutions in the past with EE and ran into that issue. Also with cookie blocking issues.

We really need more granular control over the display language, we work a lot with English sites that have Spanish translations available. 

- Sometimes the Spanish content might be on a fully/mostly translated page
https://rootedhomes.org/es/events

- Other times the Spanish content needs to display when the rest of the page is in English, like this event
https://proudground.org/calendar/filter-by/event/sesion-de-informacion-sobre-nuevo-proceso-de-aplicacion-julio_18/472/

That's why I really need to be able to just add a parameter for the language right next to the format parameter within each individual date-related tag.

Setting a cookie for the whole page, or setting a tag for the whole template is too broad an approach for the work we're doing.

For all the rest of the site's content and navigation, we just create two fluid fields, with twin fields, one for each language, then we give them all the same short name with the addition of a language postfix:

text_area_en
text_area_es

The ONLY fields we're aware of that we really can't control the display language in this way are the date fields.

<img width="997" alt="Screenshot 2023-07-06 at 8 43 30 AM" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/6145078/a46759a1-4778-48da-b809-be4106ff496c">

**Comment by intoeetive** [comment]
Pull request: https://github.com/ExpressionEngine/ExpressionEngine/pull/4669
